### PR TITLE
PyPy-ready IPython.lib.pretty, new tox.ini and consistent dict/mapping proxies pretty printing

### LINKED ISSUE #9776: Pretty printing PyPy dicts not as "dictproxy"
I'm using IPython pretty printer in my plugin for running doctests with a custom representation printer. It actually works fine in CPython, but it prints all PyPy dicts as proxies.

In CPython 2.7.12 a dict isn't a dictproxy, and it's well known that they appear with curly brackets:

``` python
In [1]: {"this": 1, "is": "a", "test": None}
Out[1]: {'is': 'a', 'test': None, 'this': 1}

In [2]: import types

In [3]: types.DictProxyType is dict
Out[3]: False
```

But in PyPy 5.3.1 (Python 2.7.10), a dict appears as a dictproxy, and these types are actually the same w.r.t. the types module:

``` python
In [1]: {"this": 1, "is": "a", "test": None}
Out[1]: <dictproxy {'is': 'a', 'test': None, 'this': 1}>

In [2]: import types

In [3]: types.DictProxyType is dict
Out[3]: True

In [4]: dict.__name__
Out[4]: 'dict'

In [5]: repr(_1)
Out[5]: "{'this': 1, 'is': 'a', 'test': None}"
```

The last outputs suggests that the assignment to ``_type_pprinters[types.DictProxyType]`` in IPython.lib.pretty replaces the dict printer. Shouldn't that assignment be done with the setdefault method instead?

### LINKED ISSUE #9821: Pretty printing types.MappingProxyType
[Since Python 3.3](https://docs.python.org/3/library/types.html#types.MappingProxyType), `types.MappingProxyType` replaces the old `types.DictProxyType` from Python 2, but now it can be instantiated. However, pretty printing it is showing only its unsorted oneliner `repr(obj)` result.

```python
In [1]: import types, string

In [2]: types.MappingProxyType({k: k for k in string.ascii_letters})
Out[2]: mappingproxy({'W': 'W', 'L': 'L', 'f': 'f', 'o': 'o', 'p': 'p', 'J': 'J', 'k': 'k', 'r': 'r', 'M': 'M', 'b': 'b', 'u': 'u', 'U': 'U', 'S': 'S', 'w': 'w', 's': 's', 'G': 'G', 'l': 'l', 'I': 'I', 'R': 'R', 'H': 'H', 'e': 'e', 'g': 'g', 'K': 'K', 'v': 'v', 't': 't', 'B': 'B', 'x': 'x', 'n': 'n', 'Y': 'Y', 'O': 'O', 'c': 'c', 'F': 'F', 'z': 'z', 'd': 'd', 'a': 'a', 'q': 'q', 'm': 'm', 'Q': 'Q', 'j': 'j', 'h': 'h', 'X': 'X', 'i': 'i', 'P': 'P', 'V': 'V', 'N': 'N', 'E': 'E', 'y': 'y', 'D': 'D', 'Z': 'Z', 'C': 'C', 'A': 'A', 'T': 'T'})
```

Tested in Python 3.5.2, IPython 5.0.0. I'm fixing this one myself due to #9776

### PR DESCRIPTION
**Pull request by danilobellini**
- Fix #9821  (Python 3 only: add a `types.MappingProxyType` pretty printer + tests)
- Fix #9776  (In PyPy, avoid replacing the `dict` representation formatter/printer by the `types.DictProxyType` one)
- Add a new `tox.ini`, easier to test multiple environments (e.g. `tox -e py36,pypy -- lib utils` call "iptest lib utils" in CPython 3.6 and PyPy) including matplotlib (but in PyPy)
- Update `.travis.yml` to test with PyPy (still known to fail, so it's in `allow_failures`)
- Fixed everything needed to ensure that `test_pretty.py` tests passes in every environment (including PyPy)
- Fixed pretty printing `super(...)` objects in PyPy. Previously, it was like:

``` python
In [1]: class A(object): pass

In [2]: super(A)
Out[2]: ---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
[...]
--> 634     p.pretty(obj.__self__)
[...]
AttributeError: 'super' object has no attribute '__self__'
```
- (Python 2 only) Add consistency to pretty print bogus named classes with `"<unknown type>"`. This can be further enhanced to get the definition name, as it's stored at least in the object `__repr__`
- Fixed `IPython.lib.pretty._dict_pprinter_factory` indentation size to be equal to the prefix size
- (CPython 2 only) Consistent pretty printing for `types.DictProxyType` with an extra `ctypes`-based test to instantiate it directly with the Python API. The pretty printing uses the format used by dict proxies `types.DictProxyType.__repr__` in CPython 2.7, printing it as `"dict_proxy({...})"`, but with the sorting and line-breaking pretty printing features.

### CODE REVIEW

**Comment by danilobellini** [comment]
Done. If required, I can rebase/stash/split the commits in this PR while it's just an experimental branch (perhaps to have a separate PR for the MappingProxyType update, or to join the `.travis.yml` updates into a single commit).

For older PyPy (Python 2.7) versions (e.g. v4.0.1), this PR has more failures, including some in `test_pretty.py`. I'm using PyPy v5.3.1 (the last stable release) but Travis CI still has the v2.5.0 version, which isn't even the last v2.x version, so I've add a PyPy installation to `.travis.yml` to ensure the results in Travis CI are the same I get here, or at least as similar as possible.

As of today, the last stable PyPy versions are v2.6.1, v4.0.1 and v5.3.1. The last PyPy3 versions are v2.4.0 (a Python 3.2 that allows u-prefixed unicode strings) and v5.2.0-alpha1 (Python 3.3), but this PR is only about PyPy, not PyPy3.

**Comment by Carreau** [comment]
Thanks a lot for that ! I'm assigning myself to look at it this week ! 🎊

**Comment by minrk** [comment]
That's great, thanks @danilobellini!

File: IPython/utils/py3compat.py [hunk 1]
```diff
@@ -292,6 +292,7 @@ def execfile(fname, glob=None, loc=None, compiler=None):
 
 
 PY2 = not PY3
+PYPY = any(k.startswith("pypy") for k in dir(sys))
```
Comment (minrk) [inline IPython/utils/py3compat.py]: I think `platform.python_implementation()` is a nicer place to look for PyPy

**Comment by Carreau** [comment]
> Done. If required, I can rebase/stash/split the commits in this PR while it's just an experimental branch (perhaps to have a separate PR for the MappingProxyType update, or to join the .travis.yml updates into a single commit).

No need to do that, the only thing we tend to avoid is to merge master into the pull requests as we have some tooling that assume merge commits are merged PR to autolist list of PRs. 

> For older PyPy (Python 2.7) versions (e.g. v4.0.1), this PR has more failures, including some in test_pretty.py. I'm using PyPy v5.3.1 (the last stable release) but Travis CI still has the v2.5.0 version, which isn't even the last v2.x version, so I've add a PyPy installation to .travis.yml to ensure the results in Travis CI are the same I get here, or at least as similar as possible.

No problem, many tests are doctest, I think we could just skip them. 

> I think platform.python_implementation() is a nicer place to look for PyPy

Good point, once the PYPY variable is moved I"m og merging this in,  and we can refine in subsequent tests !

Thanks a lot !

**Comment by minrk** [comment]
Thanks a lot!

**Comment by danilobellini** [comment]
You're welcome =)

The *test_pretty.py* is passing in PyPy 5.3.1, but in PyPy 4.0.1 there are 2 failures. I was trying to fix these, where to test in a specific PyPy version here I was using a prepared docker image: https://hub.docker.com/r/library/pypy/

One of such failures is easy to fix (the `collections.defaultdict.__module__` is `"_collections"`, so the deferred printers list just need to include a printer to `("_collections", "defaultdict")`. But the other failure would require some deeper changes, as `super()` objects didn't have the `__thisclass__` attribute in older PyPy versions: https://bitbucket.org/pypy/pypy/issues/2251

**Pull request merged**